### ISSUE TITLE
Hotspot Companion App

### ISSUE THREAD

**Issue by Popolechien** [offspot/overview#3]
⬇️

**Comment by rgaudin** [comment]
⚠️ **THIS PROJECT IS NOT COMPLETELY DEFINED YET**.

We are still in the process of defining it and the bounds of the GSoC projects are not fixed yet. Please be patient and ask for clarifications.

## Main Objective

A Mobile App to synchronize data between a running Hotspot and an Internet-reachable Kiwix Service

## Target Users

Hotspot Managers: person responsible for one or multiple Kiwix Hotspot deployment. They hold the _Admin_ credentials to said hotspot

## Workflow (draft)

Hotspot Manager:

- Installs the App on its phone/tablet
- Connects to Hotspot WiFi
- Launchs App
- Create Hotspot profile, saving its credentials
- Select Hotspot tab (auto selecting profile because there's just one profile)
- Displays:
  - List of its data packages (with size) available on Hotspot
  - List of data on phone that can be sent to this Hotspot (either data **for this hotspot** or **for any Hotspot**)
- Ticks the ones he's interested in
- Hits Sync to transfer selected Packages to/from phone. Progress bar shows.
- ➡️ at this point, communication with Hotspot is over
- Gets an internet connection (at any future point in time and location)
  - directly if Hotspot is connected to Internet
  - by enabling mobile data
  - by connecting to a different WiFi network with Internet connectivity
- Select Cloud tab
- Requests online credentials and metadata for profiles that have not been synchronized
- List all Hotspot data and size (from one or multiple hotspot) stored on the phone which have not yet been uploaded to the Cloud.
- List all data that can be downloaded into phone for all profiles's Hotspots (with size)
- Tick the one he wants (everything's ticked by default)
- Hits Sync to Upload to Cloud and Dowload to phone
- ➡️ at this point, communication with Cloud is over.

## Features

- data can flow in two ways: from cloud to hotspot and from Hotspot to cloud
- transfers are asynchronous: in most use cases, the companion is connected **either** to the Hotspot or the Internet at one time. It can be connected to both though.
- Connection to Hotspot is done via WiFi only
- App needs little knowledge about data it transfers.
- Target service URL is configurable in settings

## Scope of Work

- Sync API app for Hotspot
  - Authentication reusing existing admin credentials
  - `GET` Walks a filesystem tree for files to expose them, filtering out those created before a passed date (last known completed sync – optional). Exposes free disk space
  - `POST`: accepts file upload and writes them to a know location based on type
  - Transfer of files
- Sync API app for Cloud
  - Users DB
  - Authentication via users DB
  - `GET`: Walks a filesystem tree for files to expose them, filtering out those created before a passed date (last known completed sync – optional). Files exposed are per-hotspot and all-hotspot.
 - `POST`: accepts file upload and writes them to a know location based on hotspot and type
- Mobile App (see workflow)
- Hotspot data provider for Metrics Hotspot app (out of scope for GSoC)
- Metrics data consumer for Cloud (out of scope for GSoC)

## Unknowns

- How do we identify Hotspots?
- Mobile App technology
- File transfer technology: HTTP/FTP/rsync (protocol must have good support on mobile platforms)

**Comment by AyushGupta11442** [comment]
hello @rgaudin @kelson42 @benoit74 @Popolechien 

My name is Ayush Gupta, and I am currently pursuing a Bachelor of Technology degree from SRM University.
I bring with me a strong background in Flutter development and Dart, with nearly three years of experience. Additionally, I am proficient in React, JavaScript, and have knowledge of machine learning concepts.
Given my expertise and enthusiasm for this project, I am eager to contribute and make a meaningful impact. If you have any specific requirements or tasks in mind, I would be grateful for the opportunity to discuss them further.

**Comment by Popolechien** [comment]
Hi @AyushGupta11442 as you can see the project isn't entirely scoped yet. I'd suggest you poke around the repo (start with [good first issues](https://github.com/search?q=org%3Aoffspot+label%3A%22good+first+issue%22+is%3Aopen&type=issues) here and on /kiwix or /openzim as we won't consider project proposals from people who haven't submitted a few PRs beforehand.

**Comment by aryanA101a** [comment]
> Gets an internet connection (at any future point in time and location)

Do we have to bridge the internet connection of the mobile device to the raspberry pi to facilitate internet to other connected devices?

> Requests online credentials and metadata for profiles that have not been synchronized.

Please provide some context on this point. What are profiles here and for what they are used?

**Comment by rgaudin** [comment]
> Do we have to bridge the internet connection of the mobile device to the raspberry pi to facilitate internet to other connected devices?

No. The app will just connect to the Pi via HTTP on the Hotspot WiFi

> Please provide some context on this point. What are profiles here and for what they are used?

The App will allow one person to Sync data to/from multiple Pi with the Online service. It's not clear at this stage if the App user will have a single set of credentials and be able to _attach_ to multiple Pi for instance.
We need a way to make sure that:
- data is uploaded for a particular Pi
- person uploading data for that Pi has the right to do so

**Comment by aryanA101a** [comment]
> Create Hotspot profile, saving its credentials

What does the credentials include?
User and Password of the pi or there's more to it?

**Comment by aryanA101a** [comment]
> Displays:
> List of its data packages (with size) available on Hotspot
> List of data on phone that can be sent to this Hotspot (either data for this hotspot or for any Hotspot)

By **data package** or  **data** are we referring to zim files?

**Comment by aryanA101a** [comment]
> Sync API app for Cloud
> Users DB
> Authentication via users DB
> GET: Walks a filesystem tree for files to expose them, filtering out those created before a passed date (last known completed sync – optional). Files exposed are per-hotspot and all-hotspot.

Please provide context.
What's the use of this?

**Comment by rgaudin** [comment]
All, thank you for your comments ; I've updated the project description. It should now be somewhat clear.
We'll add the API by the end of the month so you know exactly what you'll be dealing with but it's gonna be simple: list packages, download them, mark them downloaded, etc.

**Comment by aryanA101a** [comment]
What's the need to store Hostpot credentials in the user profile?

**Comment by aryanA101a** [comment]
> > What's the need to store Hostpot credentials in the user profile?
> 
> No User profile is mentioned. We want to create a _Profile_ (naming might change) for each Hotspot the user has access to. Because hotspot API require credentials (username, password), storing them on the phone means it will be typed only once and then remembered

From Hostpot credentials I inferred the ssid and password of the wifi ap.
Okay, so you meant existing admin credentials to the hotspot.

**Comment by rgaudin** [comment]
We don't want to manage  the WiFi connection from the App. They will connect to the network first then use the app.

We want the app to be as dumb and limited as possible ; to reduce maintenance and because an app in the wild is harder to update.

**Comment by aryanA101a** [comment]
> List of its Data Packages (with size) available on Hotspot

Whats the scope of the data packages we are looking at? Location wise?
Is it limited to the data in the /data/contents/zims.

> List of Data Packages on phone that can be sent to this Hotspot
either data for this hotspot or for any Hotspot

I think this requires an asyncrounous .zim file specific search on the internal storage of the phone plus caching of results.

**Comment by rgaudin** [comment]
@aryanA101a please read the description carefully. You should not care at all about how stuff are stored on the Pi. Your app will talk to an HTTP API to get a list of items and you'll retrieve content for those items through this API.

> I think this requires an asyncrounous .zim file specific search on the internal storage of the phone plus caching of results.

I don't understand this sentence. The app won't be looking at stuff on the phone outside its working directory.

**Comment by benoit74** [comment]
I've also updated the first comment descriptions with:
- More details about Hotspot credentials
- More details about Wifi connection to the Hotspot
- Data is a single file, agnostic of any format (could be a zip, tar, zim ...)
- Total volume of data stored in the phone at a given point in time could be significant (think tens of gigabyte)
- Encryption of sensitive data (credentials) on phone storage is a nice-to-have feature
- Many small fixes

**Comment by aryanA101a** [comment]
> @aryanA101a please read the description carefully. You should not care at all about how stuff are stored on the Pi. Your app will talk to an HTTP API to get a list of items and you'll retrieve content for those items through this API.
> 
> > I think this requires an asyncrounous .zim file specific search on the internal storage of the phone plus caching of results.
> 
> I don't understand this sentence. The app won't be looking at stuff on the phone outside its working directory.

I assumed an extended scope when looking at the packages on the phone, that's why I was trying to clarify.

**Comment by anirbanmajumder0** [comment]
Hello @rgaudin @kelson42 @benoit74 @Popolechien Considering the way the hotspot is build, which type of API—SOAP or REST—will the app talk to?

**Comment by rgaudin** [comment]
HTTP REST

**Comment by aryanA101a** [comment]
Is unit testing in the scope of this project?

**Comment by rgaudin** [comment]
Up to you

**Comment by technosrijan** [comment]
@rgaudin Sir, I wanted some clarification regarding whether a GSoC candidate has been selected for this in 2024, as I checked their list of contributors website and couldn't find this project over there.

**Comment by kelson42** [comment]
@technosrijan I confirm there is no GSoC student working on this issue currently.

**Comment by Bishoywadea** [comment]
@kelson42  @rgaudin 
I'm Bishoy Wadea, a third-year Computer Engineering student and a former Microsoft intern. I'm also willing to develop a prototype for this idea to further prove my commitment. Could you guide me on how to communicate with the mentor?
Looking forward to your response!

**Comment by dev-shetty** [comment]
I went through the GSoC problem statements, and I am interested in the project on developing a mobile app bridge sync between hotspots and the cloud.

I understood the overall idea of the project, but I had a few questions

- **The ticket mentions when the user launches the app, the sync should take place.**
  - Can we have the metric data transfer from the hotspot to the mobile happening in the background when they are connected, would have a more realtime experience and can save resource/time piling up? Thoughts?


- **When the hotspot is connected to the internet can we have it directly sync with the cloud? The trip to mobile seems redundant as it is only a proxy.**
  - Is it because we don't want to touch the current implementation of the hotspot?

- **What about going with React-Native for the UI rather than Flutter?**
  - It compiles down to native components.
  - It supports cross-platform.
  - It won't add unnecessary overhead like Ionic or other WebView framework
  - Has a rich package ecosystem, hot reload, is actively maintained, and very stable.
  - Using JavaScript might be easier for new contributors to pick up, rather than having a completely new language codebase in our ecosystem.

Looking forward to your feedback.

**Comment by rgaudin** [comment]
> Can we have the metric data transfer from the hotspot to the mobile happening in the background when they are connected, would have a more realtime experience and can save resource/time piling up? Thoughts?

One of the key aspect is the control of the (mobile) user over what's being transferred. In places where this app will be useful, data cost are likely expensive. Transparency is important but that shouldn't prevent users from using their phones during transferts.

> When the hotspot is connected to the internet can we have it directly sync with the cloud? The trip to mobile seems redundant as it is only a proxy.
> Is it because we don't want to touch the current implementation of the hotspot?

That's two implementations so twice the development and maintenance. Also, it means two ways of doing the same stuff based on changing conditions. We may ultimately do it but at this stage what we need is the companion.

> What about going with React-Native for the UI rather than Flutter?

Feel free to argue for it in your proposal but our key requirements are written above: maintainability is number 1 but performance is important. This app will likely be installed on $40 android phone

**Comment by rgaudin** [comment]
The _Cloud service_ here means a custom, to-be designed/written/deployed service to centralize this. When we decide to pursue with this project, we'll document that server-side API
